### PR TITLE
Add my real-time logger code (resolves #359)

### LINKED ISSUE #359: Real-time logging to master
I want to see some output from my jobs as they are running, so I can get an idea of what's up with them and guess at whether (and when) they are going to finish. `--logDebug` does not appear to accomplish this.

It would be nice if this also worked over the network somehow.

### PR DESCRIPTION
**Pull request by adamnovak**
This isn't really done yet, and it doesn't have any tests to verify that it actually works, but with this, you ought to be able to:

```
from toil.realtimeLogger import RealtimeLogger

RealtimeLogger.get().info("This goes straight to master")
```

Probably will address #359, even though this doesn't use TCP.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Thank you for the PR. This will need a few changes. Are you throwing this over the fence or do you want to keep contributing? I am perfectly fine with either answer, it just affects our internal planning and resource allocation.

File: src/toil/job.py [hunk 1]
```diff
@@ -362,6 +363,9 @@ def startToil(job, options):
     failed jobs
             """
             setLoggingFromOptions(options)
+            # Set up real-time log message reception, and put the details in the
+            # environment to be sent out to workers.
+            RealtimeLogger.start_master()
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This should be made optional via a option. The port should also be configurable (if it isn't already).

Comment (hannes-ucsc) [inline src/toil/job.py]: And by option I mean command line option.

File: src/toil/realtimeLogger.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        """
+        Start up the master server and put its details into the options
+        namespace.
+        
+        Python logging should have already been configured.
+        
+        """
+        
+        # Start up the logging server
+        cls.logging_server = SocketServer.ThreadingUDPServer(("0.0.0.0", 0),
```
Comment (hannes-ucsc) [inline src/toil/realtimeLogger.py]: Ah, I see. Ephemeral port. That should work. Disregards my above comment about configuring a port.

File: src/toil/realtimeLogger.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        
+        Note that if the master logs here, you will see the message twice,
+        since it still goes to the normal log handlers too.
+        """
+        
+        if cls.logger is None:
+            # Only do the setup once, so we don't add a handler every time we
+            # log
+            cls.logger = logging.getLogger('realtime')
+            cls.logger.setLevel(logging.INFO)
```
Comment (hannes-ucsc) [inline src/toil/realtimeLogger.py]: Should the level be configurable?

File: src/toil/realtimeLogger.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        # Set up a thread to do all the serving in the background and exit when
+        # we do
+        cls.server_thread = threading.Thread(
+            target=cls.logging_server.serve_forever)
+        cls.server_thread.daemon = True
+        cls.server_thread.start()
+        
+        # Set options for logging in the class and the options namespace
+        # Save them in the environment so they get sent out to jobs
+        os.environ["RT_LOGGING_HOST"] = socket.getfqdn()
```
Comment (hannes-ucsc) [inline src/toil/realtimeLogger.py]: Every environment variable introduced by Toil should be prefixed with `TOIL_`. So I think `TOIL_RT_LOG_HOST` would be more appropriate.

**Comment by hannes-ucsc** [comment]
Also, UDP packets can and will arrive in any order. The kernel will also drop the if too many arrive at a time. I've observed both. As currently written, the while loop in the server would go haywire if any of those two things happened. While I don't think we should reimplement TCP, we should recover from dropped or out of order packets and print a warning if either of those happen. This is typically solved by adding a sequence number. Alternatively, let's just use TCP.

Also, I think it is safe to assume that you get an entire datagram with every call to read or at least with socket.recvfrom. So the length prefixing of packets isn't really necessary for UDP.

**Comment by adamnovak** [comment]
I can keep working on this when I can find the time. Your suggestions all make sense to me, except the out-of-order packet handling. Since you may have many workers, and they may be overloading the master with log messages, I don't want to use TCP. TCP forces the master to keep an open connection to every worker, can potentially bog down the network retransmitting dropped packets, and, if the master still can't keep up, will eventually make the workers block writing log messages if their buffers fill up. What I want is for logging messages to be immediately and unreliably fired off to the master, and for the master to just display them as they come in and silently drop them if it can't keep up.

Seeing messages out of order could potentially be confusing, but I don't think it's a good idea for a "real time" logging system to wait around overly long to put them back in order, or to wait around while requesting retransmission of log messages that dropped. Maybe I could put in a second or so of delay and some sequence numbers, to let it fix up minor ordering issues. On the other hand, that might take a disproportionate amount of code.

The only time that out-of-order or dropped datagrams can *break* the code that I have now is if a message manages to be split over multiple datagrams. What I think I should do is (somehow) work out the max packet size, keep all messages below that size, and remove my multiple-read-call loop. Logging messages shouldn't be kilobytes long; the ones that are can be either split (and reassembled out of order) or maybe truncated.

**Comment by hannes-ucsc** [comment]
Completely agree about not using TCP. This should be a best effort only. I also don't want to reimplement TCP. So no reassembly, reordering or retransmissions. But I *think* that the main server loop will get out of sync if packets are truncated, dropped or swapped and it will never recover. We should talk in person.

**Comment by adamnovak** [comment]
OK, I have addressed the potential issues with the server getting out of sync with the messages. It turns out that SocketServer just delivers every datagram to a separate handler instance. I've switched to just putting bare JSON in the datagram and not even using the stream-based interface to read it, but instead just taking the datagram contents as a string on the server side.

I also prefixed the environment variables used for communicating logging server info to the workers.

Finally, I've added a unit test, and sugared up the syntax so you can just do `RealtimeLogger.info("message")`.

**Comment by hannes-ucsc** [comment]
Continued in #633

**Pull request closed without merging**